addons/crm_partner_assign/models/crm_partner_assign.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from openerp import api, fields, models, _
+
+
+class ResPartnerGrade(models.Model):
+    _order = 'sequence'
+    _name = 'res.partner.grade'
+
+    sequence = fields.Integer()
+    active = fields.Boolean(default=True)
+    name = fields.Char(string='Level Name')
+    partner_weight = fields.Integer(string='Level Weight', default=1, help="Gives the probability to assign a lead to this partner. (0 means no assignation.)")
+
+
+class ResPartnerActivation(models.Model):
+    _name = 'res.partner.activation'
+    _order = 'sequence'
+
+    sequence = fields.Integer()
+    name = fields.Char(required=True)
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+    
+    partner_weight = fields.Integer(string='Level Weight', help="Gives the probability to assign a lead to this partner. (0 means no assignation.)")
+    grade_id = fields.Many2one('res.partner.grade', string='Level')
+    activation = fields.Many2one('res.partner.activation', index=True)
+    date_partnership = fields.Date(string='Partnership Date')
+    date_review = fields.Date(string='Latest Partner Review')
+    date_review_next = fields.Date(string='Next Partner Review')
+    # customer implementation
+    assigned_partner_id = fields.Many2one('res.partner', string='Implemented by',)
+    implemented_partner_ids = fields.One2many('res.partner', 'assigned_partner_id', string='Implementation References',)
+ 
+    @api.onchange('grade_id')
+    def onchange_grade_id(self):
+        self.partner_weight = self.grade_id.partner_weight
+
+    def search_geo_localize_partner(self, latitude, longitude, country):
+        # 1. first way: in the same country, small area
+        Partners = self.search([
+            ('partner_weight', '>', 0),
+            ('partner_latitude', '>', latitude - 2),
+            ('partner_latitude', '<', latitude + 2),
+            ('partner_longitude', '>', longitude - 1.5),
+            ('partner_longitude', '<', longitude + 1.5),
+            ('country_id', '=', country.id),
+        ])
+
```

Comment:
this function can be improved. first make common search for country and partner_weight >0 then use filtered for second and third case.